# Data Analysis with Python

## Test Bowler Analysis
Prepared By: Suman Biswas
<br>         Scientific Officer (Statistics), BARI, Gazipur, BD


## Homework 01

In [288]:
import pandas as pd
import numpy as np

#### 01. Data import

In [289]:
# Import DataFrame
df=pd.read_excel('test_cricket.xlsx', sheet_name='wickets')

#### 02. Display data

In [290]:
# Display the first 10 rows of the dataframe
display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   8/71   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   7/42   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617  10/74   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   8/24   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   8/15   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   7/37   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   7/51   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   9/83   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433  9/127   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  29.65  2.69  65.9  35   8  
3  14/149  26.67  2.83  56.3  30   3  
4   10/27  21.64  2.49  51.9  29   3  
5  11/121  27.72  2.92  56.8  18   3  
6   13/55  24.44  2.53  57.8  22   3  
7   11/60  22.95  3.24  42.3  26   5  
8  11/146  29.64  2.78  63.9  23   2  
9  14/184  28.07  2.80  60.0  34   9

#### 03. Explanation the meaning of each columns

In [291]:
print(df.columns)

Index(['Player',   'Span',    'Mat',   'Inns',  'Balls',   'Runs',   'Wkts',
          'BBI',    'BBM',    'Ave',   'Econ',     'SR',        5,       10],
      dtype='object')


Each column name can be described as:
<br> [01] Name: Name of the Player
<br> [02] Span: Duration of the Career of each player
<br> [03] Mat : No. of matches played
<br> [04] Inns: No. of innings played
<br> [05] Balls: No of balls bowled
<br> [06] Runs: Total of the runs conceded by a bowler
<br> [07] Wkts: No of total wickets taken
<br> [08] BBI: Best bowling in an Innings
<br> [09] BBM: Best bowling in a Match
<br> [10] Ave: Bowling_Average
<br> [11] Econ: Average runs conceded per over
<br> [12] SR: Strike Rate
<br> [13] 5: No of times 5 wickets taken
<br> [14] 10: No of times 10 wickets taken

#### 04. Number of rows and columns in the dataframe

In [292]:
# No. of rows
print('No. of rows: ', df.shape[0])

# No. of columns
print('No. of columns: ', df.shape[1])

No. of rows:  79
No. of columns:  14


#### 05. Data statistics and checking data types

In [293]:
# Data Statistics
display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

In [294]:
# Data types of each column
display(df.dtypes)

Player     object
Span       object
Mat         int64
Inns        int64
Balls       int64
Runs        int64
Wkts        int64
BBI        object
BBM        object
Ave       float64
Econ      float64
SR        float64
5           int64
10          int64
dtype: object

#### 06. Checking missing values

In [295]:
# Checking for missing values
print(df.isnull().sum())

Player    0
Span      0
Mat       0
Inns      0
Balls     0
Runs      0
Wkts      0
BBI       0
BBM       0
Ave       0
Econ      0
SR        0
5         0
10        0
dtype: int64


There is no missing values in the dataset

#### 07. Rename the column names

In [296]:
df_new = df.rename(columns = {'Player':'Player_Name',
                          'Span':'Duration_of_Career',
                          'Mat':'No_of_match_played', 
                          'Inns':'No_of_innings', 
                          'Balls':'No_of_balls_bowled',
                          'Runs': 'Total_Runs_conceded', 
                          'Wkts':'Total_wickets_taken', 
                          'BBI': 'Best_Bowling_in_an_Innings', 
                          'BBM':'Best_Bowling_in_a_Match',
                          'Ave':'Bowling_Average',
                          'Econ':'average_runs_conceded_per_over', 
                          'SR':'Strike_rate',
                          5: 'No_of_times_5_wickets_taken', 
                          10:'No_of_times_10_wickets_taken'})
display(df_new.head())

Player_Name Duration_of_Career  No_of_match_played  \
0  M Muralitharan (ICC/SL)          1992-2010                 133   
1           SK Warne (AUS)          1992-2007                 145   
2         A Kumble (INDIA)          1990-2008                 132   
3        JM Anderson (ENG)          2003-2021                 162   
4         GD McGrath (AUS)          1993-2007                 124   

   No_of_innings  No_of_balls_bowled  Total_Runs_conceded  \
0            230               44039                18180   
1            273               40705                17995   
2            236               40850                18355   
3            301               34791                16457   
4            243               29248                12186   

   Total_wickets_taken Best_Bowling_in_an_Innings Best_Bowling_in_a_Match  \
0                  800                       9/51                  16/220   
1                  708                       8/71                  12/128   
2                  619                       7/42                   11/71   
3                  617                      10/74                  14/149   
4                  563                       8/24                   10/27   

   Bowling_Average  average_runs_conceded_per_over  Strike_rate  \
0            22.72                            2.47         55.0   
1            25.41                            2.65         57.4   
2            29.65                            2.69         65.9   
3            26.67                            2.83         56.3   
4            21.64                            2.49         51.9   

   No_of_times_5_wickets_taken  No_of_times_10_wickets_taken  
0                           67                            22  
1                           37                            10  
2                           35                             8  
3                           30                             3  
4                           29                             3

#### 08. Removing a column from the dataframe

In [297]:
df_new = df_new.drop('Duration_of_Career', axis=1)
display(df_new.head())

Player_Name  No_of_match_played  No_of_innings  \
0  M Muralitharan (ICC/SL)                 133            230   
1           SK Warne (AUS)                 145            273   
2         A Kumble (INDIA)                 132            236   
3        JM Anderson (ENG)                 162            301   
4         GD McGrath (AUS)                 124            243   

   No_of_balls_bowled  Total_Runs_conceded  Total_wickets_taken  \
0               44039                18180                  800   
1               40705                17995                  708   
2               40850                18355                  619   
3               34791                16457                  617   
4               29248                12186                  563   

  Best_Bowling_in_an_Innings Best_Bowling_in_a_Match  Bowling_Average  \
0                       9/51                  16/220            22.72   
1                       8/71                  12/128            25.41   
2                       7/42                   11/71            29.65   
3                      10/74                  14/149            26.67   
4                       8/24                   10/27            21.64   

   average_runs_conceded_per_over  Strike_rate  No_of_times_5_wickets_taken  \
0                            2.47         55.0                           67   
1                            2.65         57.4                           37   
2                            2.69         65.9                           35   
3                            2.83         56.3                           30   
4                            2.49         51.9                           29   

   No_of_times_10_wickets_taken  
0                            22  
1                            10  
2                             8  
3                             3  
4                             3

## Homework 02

#### 01. Removing the columns BBI and BBM.

In [298]:
data=df.copy()
data=data.drop(['BBI', 'BBM'],  axis=1) # Drop 'BBI' and 'BBM'
display(data.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts    Ave  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800  22.72   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708  25.41   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619  29.65   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617  26.67   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563  21.64   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523  27.72   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519  24.44   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439  22.95   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434  29.64   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433  28.07   

   Econ    SR   5  10  
0  2.47  55.0  67  22  
1  2.65  57.4  37  10  
2  2.69  65.9  35   8  
3  2.83  56.3  30   3  
4  2.49  51.9  29   3  
5  2.92  56.8  18   3  
6  2.53  57.8  22   3  
7  3.24  42.3  26   5  
8  2.78  63.9  23   2  
9  2.80  60.0  34   9

#### 02. How many players played for ICC?

In [299]:
# splitting the 'Player' column to get the information about 'Country'
data_player = data['Player'].str.split("(", expand=True)

# concatinating 'Country' with the main dataframe
data = pd.concat([data, data_player], axis=1)

# dropping the 'Player' columns
data = data.drop('Player', axis=1)

# renaming the column names
data = data.rename(columns={0: 'Player',
                        1: 'Country'})
new_columns=['Player','Span', 'Mat', 'Inns','Balls', 'Runs','Wkts', 'Ave','Econ',
                   'SR',5, 10,'Country']
data=data[new_columns]
# remove the ")" from the 'Country' column
data['Country'] = data['Country'].str.replace(")", "")

# create a column based on a condition or function
def icc_check(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

##Finding number of player played for ICC    
data['Played_for_ICC'] = data['Country'].apply(icc_check)
print(data['Played_for_ICC'].value_counts())

No     74
Yes     5
Name: Played_for_ICC, dtype: int64


A total 5 number of players played for ICC.

#### 03. How many different countries are present in this dataset? 

In [300]:
# Removing "ICC/" from the 'Country'
data['Country']=data['Country'].str.replace("ICC/", "")
data['Country']=data['Country'].str.replace("/ICC", "")
data['Country'].value_counts()

AUS      18
ENG      15
INDIA    10
WI        9
SA        8
NZ        7
PAK       7
SL        3
BDESH     1
ZIM       1
Name: Country, dtype: int64

10 different countries are present in the dataset.

#### 04. Which player(s) had played for the longest period of time?

In [301]:
# splitting the 'Span' column based on the "-"
data_span = data['Span'].str.split("-", expand=True)

# conccatinating the new dataframe with the main dataframe
data = pd.concat([data, data_span], axis=1)

# renaming the newly created column names
data = data.rename(columns={0: "Start Year",
                        1: "End Year"})

# removing the "Span" column
data = data.drop("Span", axis=1)

# Converting the object to int
data['Start Year'] = data['Start Year'].astype('int') 
data['End Year'] = data['End Year'].astype('int')

data['Years_Played'] = data['End Year'] - data['Start Year']

data = data.drop(['Start Year', "End Year"], axis=1)

# Highest number of years played
data.sort_values(by='Years_Played', ascending = False).head()

Player  Mat  Inns  Balls   Runs  Wkts    Ave  Econ    SR   5  10  \
21      Imran Khan    88   142  19458   8258   362  22.81  2.54  53.7  23   6   
55       GS Sobers    93   159  21599   7999   235  34.03  2.22  91.9   6   0   
9     HMRKB Herath    93   170  25993  12157   433  28.07  2.80  60.0  34   9   
0   M Muralitharan   133   230  44039  18180   800  22.72  2.47  55.0  67  22   
2         A Kumble   132   236  40850  18355   619  29.65  2.69  65.9  35   8   

   Country Played_for_ICC  Years_Played  
21     PAK             No            21  
55      WI             No            20  
9       SL             No            19  
0       SL            Yes            18  
2    INDIA             No            18

Imran Khan of Pakistan played highest (21) number of years.

#### 05. Which player(s) had played for the shortest period of time?

In [302]:
# SHortest period of time
data.sort_values(by='Years_Played', ascending = True).head()

Player  Mat  Inns  Balls  Runs  Wkts    Ave  Econ    SR   5  10  \
44      GP Swann    60   109  15349  7642   255  29.96  2.98  60.1  17   3   
71      K Rabada    47    86   8785  4846   213  22.75  3.30  41.2  10   4   
54    Yasir Shah    45    83  13601  7247   235  30.83  3.19  57.8  16   3   
61   SJ Harmison    63   115  13375  7192   226  31.82  3.22  59.1   8   1   
72  JR Hazlewood    55   103  11887  5438   212  25.65  2.74  56.0   9   0   

   Country Played_for_ICC  Years_Played  
44     ENG             No             5  
71      SA             No             6  
54     PAK             No             7  
61     ENG            Yes             7  
72     AUS             No             7

GP Swann of England played shortest (5 years) period of times.

#### 06. How many Australian Bowlers are present in this dataset?

In [303]:
data_aus=(data['Country']=='AUS') # Filter country by AUS
data_aus.value_counts()

False    61
True     18
Name: Country, dtype: int64

Total 18 Australian Bowlers are present in this dataset.

####  07. Is there any Bangladeshi player present in this dataset?

In [304]:
# Filter country data by BDESH
data[data['Country']=='BDESH']

Player  Mat  Inns  Balls  Runs  Wkts    Ave  Econ    SR   5  10  \
70  Shakib Al Hasan    58    98  13415  6679   215  31.06  2.98  62.3  18   2   

   Country Played_for_ICC  Years_Played  
70   BDESH             No            14

There is  one Bangladeshi player present in this dataset who is Shakib Al Hasan.

#### 08.Which player had the lowest economy rate?

In [305]:
# sorting by lowest Econ
data.sort_values(by='Econ', ascending = True).head()

Player  Mat  Inns  Balls  Runs  Wkts    Ave  Econ    SR   5  10  \
32      LR Gibbs    79   148  27115  8989   309  29.09  1.98  87.7  18   2   
47      R Benaud    63   116  19108  6704   248  27.03  2.10  77.0  16   1   
35  DL Underwood    86   151  21862  7674   297  25.83  2.10  73.6  17   6   
39       BS Bedi    67   118  21364  7637   266  28.71  2.14  80.3  14   1   
68   CV Grimmett    37    67  14513  5231   216  24.21  2.16  67.1  21   7   

   Country Played_for_ICC  Years_Played  
32      WI             No            18  
47     AUS             No            12  
35     ENG             No            16  
39   INDIA             No            13  
68     AUS             No            11

LR Gibbs had the lowest economy rate of 1.98 in the test cricket

#### 09.Which player had the lowest strike rate?

In [306]:
# sorting by lowest SR
data.sort_values(by='SR', ascending = True).head()

Player  Mat  Inns  Balls   Runs  Wkts    Ave  Econ    SR   5  10  \
71      K Rabada    47    86   8785   4846   213  22.75  3.30  41.2  10   4   
7       DW Steyn    93   171  18608  10077   439  22.95  3.24  42.3  26   5   
20  Waqar Younis    87   154  16224   8788   373  23.56  3.25  43.4  22   5   
19   MD Marshall    81   151  17584   7876   376  20.94  2.68  46.7  22   4   
25     AA Donald    72   129  15519   7344   330  22.25  2.83  47.0  20   3   

   Country Played_for_ICC  Years_Played  
71      SA             No             6  
7       SA             No            15  
20     PAK             No            14  
19      WI             No            13  
25      SA             No            10

K Rabada had the lowest bowling strike rate of 41.2 in the test cricket.

#### 10. Which player had the lowest bowling average?

In [307]:
# sorting by lowest Ave
data.sort_values(by='Ave', ascending = True).head()

Player  Mat  Inns  Balls   Runs  Wkts    Ave  Econ    SR   5  10  \
19  MD Marshall    81   151  17584   7876   376  20.94  2.68  46.7  22   4   
41     J Garner    58   111  13169   5433   259  20.97  2.47  50.8   7   0   
15  CEL Ambrose    98   179  22103   8501   405  20.99  2.30  54.5  22   3   
33   FS Trueman    67   127  15178   6625   307  21.57  2.61  49.4  17   3   
4    GD McGrath   124   243  29248  12186   563  21.64  2.49  51.9  29   3   

   Country Played_for_ICC  Years_Played  
19      WI             No            13  
41      WI             No            10  
15      WI             No            12  
33     ENG             No            13  
4      AUS             No            14

MD Marshall had the lowest bowling average of 20.94 in the test cricket.